### PR TITLE
feat: add `multicall2` contract

### PR DESCRIPTION
**Pull request by 0xArdi**
## Proposed changes

This PR backports the makerDAO multicall2 contract from https://github.com/valory-xyz/agent-academy-1.

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [x] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [x] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have locally run services that could be impacted and they do not present failures derived from my changes

### CODE REVIEW

**Comment by Adamantios** [comment]
There is a merge error, you can cherry pick 6278dd37faa4f57839f4deb5d49f9c44ce6ffd7d to resolve this as I also had the same :sweat_smile: .

**Review by Adamantios — CHANGES_REQUESTED** [review]
Looks good, just a merge error.

**Comment by 0xArdi** [comment]
The e2e failure is not related. The failure is on `TestRegistrationStartUpFourAgentsCatchUp` because the trigger for resetting agent0 took a bit longer than usual. Additionally, upon restarting agent0 couldnt connect to the local libp2p node.

**Review by Adamantios — APPROVED** [review]
Not a merge error after all: https://github.com/valory-xyz/open-autonomy/pull/1754#issuecomment-1406615277

File: packages/valory/contracts/multicall2/contract.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+        Implement this method in the sub class if you want
+        to handle the contract requests manually.
+
+        :param ledger_api: the ledger apis.
+        :param contract_address: the contract address.
+        :param kwargs: the keyword arguments.
+        :return: the tx  # noqa: DAR202
+        """
+        raise NotImplementedError
```
Comment (DavidMinarsch) [inline packages/valory/contracts/multicall2/contract.py:57]: Please add pragma nocovers for the unimplemented methods

**Review by Adamantios — APPROVED** [review]

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**